api/src/main/java/org/pmiops/workbench/access/AccessModuleServiceImpl.java

```diff
@@ -0,0 +1,133 @@
+              .put(BypassTimeTargetProperty.ERA_COMMONS_BYPASS_TIME, AccessModuleName.ERA_COMMONS)
+              .put(
+                  BypassTimeTargetProperty.COMPLIANCE_TRAINING_BYPASS_TIME,
+                  AccessModuleName.RT_COMPLIANCE_TRAINING)
+              .put(
+                  BypassTimeTargetProperty.TWO_FACTOR_AUTH_BYPASS_TIME,
+                  AccessModuleName.TWO_FACTOR_AUTH)
+              .put(BypassTimeTargetProperty.RAS_LINK_LOGIN_GOV, AccessModuleName.RAS_LOGIN_GOV)
+              .put(
+                  BypassTimeTargetProperty.DATA_USE_AGREEMENT_BYPASS_TIME,
+                  AccessModuleName.DATA_USER_CODE_OF_CONDUCT)
+              .build();
+
+  @Autowired
+  public AccessModuleServiceImpl(
+      Provider<List<DbAccessModule>> dbAccessModulesProvider,
+      Clock clock,
+      UserAccessModuleDao userAccessModuleDao,
+      UserDao userDao,
+      UserServiceAuditor userServiceAuditor,
+      Provider<WorkbenchConfig> configProvider) {
+    this.dbAccessModulesProvider = dbAccessModulesProvider;
+    this.clock = clock;
+    this.userAccessModuleDao = userAccessModuleDao;
+    this.userDao = userDao;
+    this.userServiceAuditor = userServiceAuditor;
+    this.configProvider = configProvider;
+  }
+
+  @Override
+  public void updateBypassTime(long userId, AccessModule accessModuleName, boolean isBypassed) {
+    final DbUser user = userDao.findUserByUserId(userId);
+    final List<DbUserAccessModule> dbUserAccessModules = userAccessModuleDao.getAllByUser(user);
+    DbAccessModule accessModule =
+        dbAccessModulesProvider.get().stream()
+            .filter(a -> a.getName() == clientAccessModuleToStorage(accessModuleName))
+            .findFirst()
+            .orElseThrow(
+                () ->
+                    new BadRequestException(
+                        "There is no access module named: " + accessModuleName.toString()));
+    if (!accessModule.getBypassable()) {
+      throw new ForbiddenException("Bypass: " + accessModuleName.toString() + " is not allowed.");
+    }
+    Optional<DbUserAccessModule> retrievedAccessModule =
+        dbUserAccessModules.stream()
+            .filter(m -> m.getAccessModule().getName().equals(accessModule.getName()))
+            .findFirst();
+    final Timestamp newBypassTime;
+    final Timestamp previousBypassTime;
+    final DbUserAccessModule userAccessModuleToUpdate;
+
+    logger.info(
+        String.format(
+            "Bypassing %s(uid: %d) for module %s",
+            user.getContactEmail(), userId, accessModule.getName()));
+    previousBypassTime = retrievedAccessModule.map(DbUserAccessModule::getBypassTime).orElse(null);
+    newBypassTime = isBypassed ? new Timestamp(clock.instant().toEpochMilli()) : null;
+    userAccessModuleToUpdate =
+        retrievedAccessModule
+            .orElseGet(() -> new DbUserAccessModule().setUser(user).setAccessModule(accessModule))
```

Comment:
we want to add a row if we are bypassing and it doesn't already exist, so this is good.

But do we want to add a row if isBypassed is false?  I am not sure.  What do you think?

Reply:
I think the logic here is simpler because we always add a row if not present.
 If we want to add the logic to not adding row when isBypassed is false, I did not see the benefit of that. 
In our following logic to check if user bypassed, we still need to load all modules check if row present or bypassTime is null. Because we have the case: Bypass then unbypass

Reply:
sgtm